### ISSUE TITLE
no service found for - "org.qt-project.qt.mediaplayer"

### ISSUE THREAD

**Issue by nnttoo** [probonopd/linuxdeployqt#148]
I tried deploy an application using QMediaPlayer to play an mp3 file, when i run using qt creator, it runs fine, but after i deploy and run in ubuntu 15.10, the file is not played, i see error message 

`"defaultServiceProvider::requestService(): no service found for - "org.qt-project.qt.mediaplayer"`

I've tried to manually copy the mediaservice from plugin folder but it does not work.

**Comment by probonopd** [comment]
Is the source to the application available, so that I could have a look?

**Comment by nnttoo** [comment]
Source code and deploy results are here:
https://www.dropbox.com/s/d2174a23ae1k2u3/QTAUdio.zip?dl=0

**Comment by probonopd** [comment]
Getting 

> Error (429)
This link is temporarily disabled. The person who shared it hit their daily limit of traffic or downloads. Learn about traffic limits. 

Can you please upload the source to GitHub?

**Comment by nnttoo** [comment]
https://github.com/nnttoo/QTAudio 
i still uploading the deploy result

**Comment by probonopd** [comment]
Please check https://github.com/probonopd/QTAudio/blob/patch-1/.travis.yml

AppImage:
https://transfer.sh/gPbCi/QTAudioTest-git.8235d7d-x86_64.AppImage

Nothing manually copied into the AppDir.

**Comment by nnttoo** [comment]
the appimage run perfectly in my linux mint, but i still can't play the audio in ubuntu 15.10 
the error now is 

`Your GStreamer installation is missing a plug-in`

**Comment by probonopd** [comment]
Which plug-in might it be missing?

**Comment by nnttoo** [comment]
I do not know about gstreamer, and i don know which plugin is missing
The application is cant play mp3, but it runs fine when playing ogg file, does qt need gstreamer to play mp3?

If all gstreamer plugins are not too large, why not just bundle  especially for important files that are often used, eg mp3, mp4 and avi.

i want to try toi add gstream file in bin folder, but i dont know where do I get the gstreamer binary file?

**Comment by TheAssassin** [comment]
Check out the contents of the `gstreamer1.0-plugins-ugly` Debian package. I found a LAME binary in there. LAME is like an open source MP3 implementation, so I think if you copy the libraries in this package to the AppImage, it could work as expected.

**Comment by nnttoo** [comment]
i have tried  copy all .so file from gstreamer-1.0 folder from gstreamer1.0-plugins-ugly_1.2.3-2build1_amd64.deb, to lib folder  but still cant play mp3

**Comment by TheAssassin** [comment]
Because GStreamer doesn't 'load' the libraries on startup using the linker (a kind of hard dependency, you'd see linker errors) but uses the dynamic linking loader with `dlopen()` (I assume at least), it probably has hard-coded search paths in the binary. Thus, it tries to find them on your system and not in the mounted AppImage. We'll have to work around this problem.
I assume you'd need to use either unionfs or binpatch, one of them should work. I'll have to test that tonight.

**Comment by probonopd** [comment]
> I assume you'd need to use either unionfs or binpatch

...or set the `GST_PLUGIN_PATH` environment variable using a launcher script. In fact, I'd be willing to consider setting this in `runtime.c` for all AppImages.

**Comment by TheAssassin** [comment]
Good idea. We could set it to `<appimage mount point>/usr/lib/gstreamer-1.0` as on the system. I do not think we need support for gstreamer0.10 at the moment, it's slowly dying and I guess people could just copy their libraries to the `-1.0` directory, too.

Maybe we could even automate the process by detecting which version is used, and copy the according plugins, like with Qt.

**Comment by probonopd** [comment]
..but what are the "according" plugins, how do we know which ones to bundle?

**Comment by TheAssassin** [comment]
That's the one million dollar question, isn't it?

I guess there's a way to search the libraries for the plugin names somehow recursively, but that's quite naive and rather inefficient. Many gstreamer plugins, like the audio library, dynamically decide whether to support a codec or not. So it's probably necessary to bundle more libraries than actually required for an application. That's probably not a problem though, I'd prefer to bundle too many libs, and have the packager optionally remove unused libraries.

Ideally, we'll find a gstreamer companion tool that recursively resolves the dependencies and gets us a list of libraries we can package.
